tests/catch/include/hip_test_rtc.hh

```diff
@@ -0,0 +1,279 @@
+  std::vector<const char*> options{};
+#ifdef __HIP_PLATFORM_AMD__
+
+  int deviceCount;
+  REQUIRE(hipSuccess == hipGetDeviceCount(&deviceCount));
+
+  std::set<std::string> architectures{};
+  for (int i = 0; i < deviceCount; ++i) {
+    hipDeviceProp_t props;
+    REQUIRE(hipSuccess == hipGetDeviceProperties(&props, i));
+    architectures.insert(std::string{"--gpu-architecture="} + props.gcnArchName);
+  }
+
+  for (auto& architecture : architectures) {
+    options.push_back(architecture.c_str());
+  }
+#else
+  options.push_back("--fmad=false");
+#endif
+
+  REQUIRE(HIPRTC_SUCCESS == hiprtcAddNameExpression(rtcProgram, kernelNameExpression.c_str()));
+  REQUIRE(HIPRTC_SUCCESS == hiprtcCompileProgram(rtcProgram, 1, options.data()));
+
+  return rtcProgram;
+}
+
+/**
+ * @brief Get a typename as a string
+ *
+ * @tparam T The typename
+ * @return std::string the string representation of T
+ */
+template <typename T> std::string getTypeName() {
+  std::string name, prefix, suffix;
+
+
+#ifdef __clang__
+  name = __PRETTY_FUNCTION__;
+  prefix = "std::string HipTest::getTypeName() [T = ";
+  suffix = "]";
+#elif defined(__GNUC__)
+  name = __PRETTY_FUNCTION__;
+  prefix = "std::string HipTest::getTypeName() [with T = ";
+  suffix = "; std::string = std::__cxx11::basic_string<char>]";
+#elif defined(_MSC_VER)
+  name = __FUNCSIG__;
+  prefix = "std::string __cdecl HipTest::getTypeName<";
+  suffix = ">(void)";
+#endif
+
+  return name.substr(prefix.size(), name.rfind(suffix) - prefix.size());
+}
+
+/**
+ * @brief Tells the user that the kernels are using HIP RTC. Prints only once per test.
+ *
+ */
+static inline void printInfo() {
+  static bool alreadyPrinted{false};
+
+  if (!alreadyPrinted) {
+    std::cout << "INFO: This test is running using HIP RTC to compile and run the kernels."
+              << std::endl;
+    alreadyPrinted = true;
+  }
+}
+
+/**
+ * @brief Compiles and launches a kernel using HIP RTC
+ *
+ * @tparam Typenames A list of typenames used by the kernel (unused if the kernel is not a
+ * template).
+ * @tparam Args A list of kernel arguments to be forwarded.
+ * @param getKernelName A function wrapper that returns the name of the kernel to launch (check
+ * kernels.hh for more info)
+ * @param numBlocks
+ * @param numThreads
+ * @param memPerBlock
+ * @param stream
+ * @param packedArgs A list of kernel arguments to be forwarded.
+ */
+template <typename... Typenames, typename... Args>
+void launchRTCKernel(std::string (*getKernelName)(), dim3 numBlocks, dim3 numThreads,
+                     std::uint32_t memPerBlock, hipStream_t stream, Args&&... packedArgs) {
+
+  printInfo();
+  TestContext& testContext = TestContext::get();
+  std::string kernelName = (*getKernelName)();
+
+  std::vector<std::string> kernelTypenames{std::string(HipTest::getTypeName<Typenames>())...};
+  std::string kernelExpression = reconstructExpression(kernelName, kernelTypenames);
+
+  if (testContext.getFunction(kernelExpression) == nullptr) {
```

Comment:
You could maybe put this logic into compileRTC so that you call it with the source and it returns the compiled kernel, without you having to check in this function whether or not it's already been compiled. But that's maybe not worth the effort.